File: ios/Core/DownloadManager.m
//
//  DownloadManager.m
//  HuaweiObs
//
//  Created by react-native-huawei-obs
//

#import "DownloadManager.h"
#import "EventEmitter.h"
#import "ObsClientHolder.h"
#import "ErrorMapper.h"
#import <OBS/OBS.h>

#define OBSLog(tag, fmt, ...) NSLog(@"[%@] " fmt, tag, ##__VA_ARGS__)
static NSString *const TAG = @"DownloadManager";

@implementation DownloadParams

- (instancetype)initWithTaskId:(NSString *)taskId
                     objectKey:(NSString *)objectKey
                        bucket:(NSString *)bucket
                      savePath:(NSString *)savePath
                         range:(nullable NSString *)range
                     versionId:(nullable NSString *)versionId {
    if (self = [super init]) {
        _taskId = [taskId copy];
        _objectKey = [objectKey copy];
        _bucket = [bucket copy];
        _savePath = [savePath copy];
        _range = [range copy];
        _versionId = [versionId copy];
        _startTimeMs = [[NSDate date] timeIntervalSince1970] * 1000;
    }
    return self;
}

@end

@implementation DownloadResult

- (instancetype)initWithTaskId:(NSString *)taskId
                     objectKey:(NSString *)objectKey
                      savePath:(NSString *)savePath
                          size:(int64_t)size
                         etag:(NSString *)etag
                      duration:(NSTimeInterval)duration
                      avgSpeed:(double)avgSpeed {
    if (self = [super init]) {
        _taskId = [taskId copy];
        _objectKey = [objectKey copy];
        _savePath = [savePath copy];
        _size = size;
        _etag = [etag copy];
        _duration = duration;
        _avgSpeed = avgSpeed;
    }
    return self;
}

@end

typedef NS_ENUM(NSInteger, DownloadTaskStatus) {
    DownloadTaskStatusPending,
    DownloadTaskStatusDownloading,
    DownloadTaskStatusCompleted,
    DownloadTaskStatusFailed,
    DownloadTaskStatusCanceled
};

@interface DownloadTaskState : NSObject
@property (nonatomic, copy) NSString *taskId;
@property (nonatomic, strong) DownloadParams *params;
@property (nonatomic, copy) NSString *status;
@property (nonatomic, assign) int64_t downloadedBytes;
@property (nonatomic, assign) int64_t totalBytes;
@property (nonatomic, copy) NSString *etag;
@property (nonatomic, strong) dispatch_block_t cancelBlock;
@property (nonatomic, strong) NSError *error;
@property (nonatomic, assign) BOOL cancelled;
@end

@implementation DownloadTaskState
@end

static const NSInteger BUFFER_SIZE = 8192;
static const NSTimeInterval PROGRESS_UPDATE_INTERVAL = 0.1; // 100ms

@interface DownloadManager ()
@property (nonatomic, strong) ObsClientHolder *obsClientHolder;
@property (nonatomic, strong) EventEmitter *eventEmitter;
@property (nonatomic, strong) NSMutableDictionary<NSString *, DownloadTaskState *> *tasks;
@property (nonatomic, strong) dispatch_queue_t taskQueue;
@property (nonatomic, strong) NSLock *lock;
@end

@implementation DownloadManager

- (instancetype)initWithObsClientHolder:(ObsClientHolder *)obsClientHolder
                           eventEmitter:(EventEmitter *)eventEmitter {
    if (self = [super init]) {
        _obsClientHolder = obsClientHolder;
        _eventEmitter = eventEmitter;
        _tasks = [NSMutableDictionary dictionary];
        _taskQueue = dispatch_queue_create("com.huaweiobs.download", DISPATCH_QUEUE_CONCURRENT);
        _lock = [[NSLock alloc] init];
    }
    return self;
}

- (NSString *)startDownloadWithParams:(DownloadParams *)params {
    OBSLog(TAG, @"[StartDownload] taskId=%@  objectKey=%@  savePath=%@",
           params.taskId, params.objectKey, params.savePath);
    DownloadTaskState *taskState = [[DownloadTaskState alloc] init];
    taskState.taskId = params.taskId;
    taskState.params = params;
    taskState.status = @"PENDING";
    taskState.downloadedBytes = 0;
    taskState.totalBytes = 0;
    taskState.etag = @"";
    taskState.cancelled = NO;
    
    [self.lock lock];
    self.tasks[params.taskId] = taskState;
    [self.lock unlock];
    
    // Execute download in background
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [self executeDownload:taskState];
    });
    
    return params.taskId;
}

- (void)executeDownload:(DownloadTaskState *)taskState {
    DownloadParams *params = taskState.params;
    NSError *error = nil;
    
    @try {
        // Check credentials expiry
        if ([self.obsClientHolder isCredentialsExpired]) {
            @throw [NSError errorWithDomain:@"DownloadManager" code:-1
                                   userInfo:@{NSLocalizedDescriptionKey: @"Credentials expired",
                                             @"code": @"E_AUTH_EXPIRED"}];
        }
        
        // Create save directory
        NSURL *savedURL = [NSURL fileURLWithPath:params.savePath];
        NSFileManager *fileManager = [NSFileManager defaultManager];
        NSURL *dirURL = [savedURL URLByDeletingLastPathComponent];
        
        if (![fileManager fileExistsAtPath:[dirURL path]]) {
            [fileManager createDirectoryAtURL:dirURL 
                  withIntermediateDirectories:YES 
                                   attributes:nil 
                                        error:nil];
        }
        
        if ([fileManager fileExistsAtPath:params.savePath]) {
            [fileManager removeItemAtPath:params.savePath error:nil];
        }
        
        // Send downloadStart event
        taskState.status = @"DOWNLOADING";
        [self.eventEmitter emitWithEventName:@"downloadStart" 
                                     params:@{
                                         @"taskId": params.taskId,
                                         @"objectKey": params.objectKey
                                     }];
        
        // Create download request
        OBSClient *client = [self.obsClientHolder getClientWithError:&error];
        if (!client) {
            @throw error ?: [NSError errorWithDomain:@"DownloadManager" code:-1
                                            userInfo:@{NSLocalizedDescriptionKey: @"OBS client not available"}];
        }
        
        OBSGetObjectToFileRequest *request = [[OBSGetObjectToFileRequest alloc]
            initWithBucketName:params.bucket
                     objectKey:params.objectKey
              downloadFilePath:params.savePath];
        if (params.range) {
            request.range = params.range;
        }
        if (params.versionId) {
            request.versionID = params.versionId;
        }
        
        // Set download progress callback
        request.downloadProgressBlock = ^(int64_t bytesWritten, int64_t totalBytesWritten, int64_t totalBytesExpectedToWrite) {
            if (taskState.cancelled) return;
            taskState.downloadedBytes = totalBytesWritten;
            if (totalBytesExpectedToWrite > 0) {
                taskState.totalBytes = totalBytesExpectedToWrite;
            }
            [self emitDownloadProgress:taskState force:NO];
        };
        
        // Execute download
        OBSBFTask *task = [client invokeRequest:request];
        [task waitUntilFinished];
        
        if (task.error) {
            @throw task.error;
        }
        
        // Check cancellation after download completes
        if (taskState.cancelled) return;
        
        OBSGetObjectResponse *response = task.result;
        taskState.etag = response.etag ?: @"";
        
        // Get file size from downloaded file
        NSError *fileError = nil;
        NSDictionary *fileAttributes = [[NSFileManager defaultManager] attributesOfItemAtPath:params.savePath error:&fileError];
        if (fileAttributes) {
            taskState.totalBytes = [fileAttributes fileSize];
            taskState.downloadedBytes = taskState.totalBytes;
        }
        
        OBSLog(TAG, @"[Download] object downloaded  size=%lld bytes  etag=%@",
               taskState.totalBytes, taskState.etag);
        
        // Send final 100% progress
        [self emitDownloadProgress:taskState force:YES];
        
        // Download complete
        NSTimeInterval duration = ([[NSDate date] timeIntervalSince1970] * 1000) - params.startTimeMs;
        double avgSpeed = duration > 0 ? (taskState.totalBytes * 1000.0) / duration : 0.0;
        
        OBSLog(TAG, @"[Download] completed  objectKey=%@  size=%lld bytes  duration=%.0fms  speed=%.1f KB/s",
               params.objectKey, taskState.totalBytes, duration, avgSpeed / 1024);
        
        taskState.status = @"COMPLETED";
        
        // Send success event
        [self.eventEmitter emitWithEventName:@"downloadSuccess" 
                                     params:@{
                                         @"taskId": params.taskId,
                                         @"objectKey": params.objectKey,
                                         @"savePath": params.savePath,
                                         @"size": @(taskState.totalBytes),
                                         @"etag": taskState.etag,
                                         @"duration": @(duration),
                                         @"avgSpeed": @(avgSpeed)
                                     }];
        
        // Clear throttle record
        [self.eventEmitter clearThrottle:params.taskId];
        
    } @catch (NSError *e) {
        if (!taskState.cancelled) {
            [self handleDownloadError:taskState error:e];
        }
    }
}

- (void)emitDownloadProgress:(DownloadTaskState *)taskState force:(BOOL)force {
    double percentage = taskState.totalBytes > 0 ? 
        (taskState.downloadedBytes * 100.0 / taskState.totalBytes) : 0.0;
    
    NSDictionary *params = @{
        @"taskId": taskState.taskId,
        @"downloadedBytes": @(taskState.downloadedBytes),
        @"totalBytes": @(taskState.totalBytes),
        @"percentage": @((NSInteger)percentage)
    };
    
    [self.eventEmitter emitProgressWithTaskId:taskState.taskId
                                   eventName:@"downloadProgress"
                                      params:params
                                       force:force];
}

- (void)handleDownloadError:(DownloadTaskState *)taskState error:(NSError *)error {
    NSDictionary *obsError = [ErrorMapper mapError:error];
    taskState.error = error;
    taskState.status = @"FAILED";
    
    OBSLog(TAG, @"[Download] failed  taskId=%@  error=%@", taskState.taskId, obsError[@"message"]);
    
    // Send error event
    [self.eventEmitter emitWithEventName:@"downloadError" 
                                 params:@{
                                     @"taskId": taskState.taskId,
                                     @"code": obsError[@"code"],
                                     @"message": obsError[@"message"],
                                     @"isRetryable": obsError[@"isRetryable"]
                                 }];
    
    // Clear throttle record
    [self.eventEmitter clearThrottle:taskState.taskId];
}

- (void)cancelDownloadWithTaskId:(NSString *)taskId error:(NSError **)error {
    [self.lock lock];
    DownloadTaskState *taskState = self.tasks[taskId];
    [self.lock unlock];
    
    if (!taskState) {
        if (error) {
            *error = [NSError errorWithDomain:@"DownloadManager" code:-1
                                     userInfo:@{NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Task not found: %@", taskId]}];
        }
        return;
    }
    
    taskState.cancelled = YES;
    taskState.status = @"CANCELED";
    [self.lock lock];
    [self.tasks removeObjectForKey:taskId];
    [self.lock unlock];
    
    // Send cancel event
    [self.eventEmitter emitWithEventName:@"downloadCancel" 
                                 params:@{@"taskId": taskId}];
}

- (void)cancelAll {
    [self.lock lock];
    NSArray *taskIds = [self.tasks.allKeys copy];
    [self.lock unlock];
    
    for (NSString *taskId in taskIds) {
        [self cancelDownloadWithTaskId:taskId error:nil];
    }
}

- (nullable NSDictionary *)getTaskStatusWithTaskId:(NSString *)taskId {
    [self.lock lock];
    DownloadTaskState *taskState = self.tasks[taskId];
    [self.lock unlock];
    
    if (!taskState) {
        return nil;
    }
    
    double percentage = taskState.totalBytes > 0 ? 
        (taskState.downloadedBytes * 100.0 / taskState.totalBytes) : 0.0;
    
    NSDictionary *progress = @{
        @"taskId": taskState.taskId,
        @"downloadedBytes": @(taskState.downloadedBytes),
        @"totalBytes": @(taskState.totalBytes),
        @"percentage": @((NSInteger)percentage)
    };
    
    NSDictionary *status = @{
        @"taskId": taskState.taskId,
        @"type": @"download",
        @"objectKey": taskState.params.objectKey,
        @"status": taskState.status,
        @"progress": progress
    };
    
    return status;
}

- (NSArray<NSDictionary *> *)getAllTasks {
    [self.lock lock];
    NSArray *taskStates = [self.tasks.allValues copy];
    [self.lock unlock];
    
    NSMutableArray *result = [NSMutableArray array];
    for (DownloadTaskState *taskState in taskStates) {
        NSDictionary *task = @{
            @"taskId": taskState.taskId,
            @"type": @"download",
            @"objectKey": taskState.params.objectKey,
            @"status": taskState.status
        };
        [result addObject:task];
    }
    return result;
}

- (void)clearCompletedTasks {
    [self.lock lock];
    NSMutableArray *keysToRemove = [NSMutableArray array];
    for (NSString *taskId in self.tasks) {
        DownloadTaskState *taskState = self.tasks[taskId];
        if ([taskState.status isEqualToString:@"COMPLETED"] ||
            [taskState.status isEqualToString:@"FAILED"] ||
            [taskState.status isEqualToString:@"CANCELED"]) {
            [keysToRemove addObject:taskId];
        }
    }
    [self.tasks removeObjectsForKeys:keysToRemove];
    [self.lock unlock];
}

- (void)destroy {
    [self cancelAll];
    [self.lock lock];
    [self.tasks removeAllObjects];
    [self.lock unlock];
}

@end

File: android/src/main/java/com/huaweiobs/core/DownloadManager.java
package com.huaweiobs.core;

import android.util.Log;

import com.huaweiobs.utils.EventEmitter;
import com.huaweiobs.utils.ErrorMapper;
import com.huaweiobs.utils.OBSException;
import com.obs.services.model.GetObjectRequest;
import com.obs.services.model.ObsObject;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Download Manager
 * Manages file download operations
 */
public class DownloadManager {
    private static final String TAG ="DownloadManager";
    private static final int BUFFER_SIZE = 8192;
    private static final long PROGRESS_UPDATE_INTERVAL = 100L; // ms

    private final ObsClientHolder obsClientHolder;
    private final EventEmitter eventEmitter;
    private final ConcurrentHashMap<String, DownloadTaskState> tasks = new ConcurrentHashMap<>();
    private final ExecutorService executorService;

    /**
     * Download parameters
     */
    public static class DownloadParams {
        public final String taskId;
        public final String objectKey;
        public final String bucket;
        public final String savePath;
        public final String range;
        public final String versionId;
        public final long startTimeMs;

        public DownloadParams(String taskId, String objectKey, String bucket, String savePath,
                              String range, String versionId) {
            this.taskId = taskId;
            this.objectKey = objectKey;
            this.bucket = bucket;
            this.savePath = savePath;
            this.range = range;
            this.versionId = versionId;
            this.startTimeMs = System.currentTimeMillis();
        }
    }

    /**
     * Download result
     */
    public static class DownloadResult {
        public final String taskId;
        public final String objectKey;
        public final String savePath;
        public final long size;
        public final String etag;
        public final long duration;
        public final double avgSpeed;

        public DownloadResult(String taskId, String objectKey, String savePath,
                              long size, String etag, long duration, double avgSpeed) {
            this.taskId = taskId;
            this.objectKey = objectKey;
            this.savePath = savePath;
            this.size = size;
            this.etag = etag;
            this.duration = duration;
            this.avgSpeed = avgSpeed;
        }
    }

    /**
     * Download task state
     */
    private static class DownloadTaskState {
        final String taskId;
        final DownloadParams params;
        String status = "PENDING";
        final AtomicLong downloadedBytes = new AtomicLong(0);
        long totalBytes = 0;
        String etag = "";
        Future<?> future;
        OBSException error;

        DownloadTaskState(String taskId, DownloadParams params) {
            this.taskId = taskId;
            this.params = params;
        }
    }

    public DownloadManager(ObsClientHolder obsClientHolder, EventEmitter eventEmitter) {
        this.obsClientHolder = obsClientHolder;
        this.eventEmitter = eventEmitter;
        this.executorService = Executors.newCachedThreadPool();
    }

    /**
     * Start download
     */
    public String startDownload(DownloadParams params) {
        Log.i(TAG, "[StartDownload] taskId=" + params.taskId + "  objectKey=" + params.objectKey + "  savePath=" + params.savePath);
        DownloadTaskState taskState = new DownloadTaskState(params.taskId, params);
        tasks.put(params.taskId, taskState);

        // Execute download in background
        taskState.future = executorService.submit(() -> {
            try {
                executeDownload(taskState);
            } catch (Exception e) {
                handleDownloadError(taskState, e);
            }
        });

        return params.taskId;
    }

    /**
     * Execute download flow
     */
    private void executeDownload(DownloadTaskState taskState) throws Exception {
        DownloadParams params = taskState.params;

        try {
            // Check credentials expiry
            if (obsClientHolder.isCredentialsExpired()) {
                throw new OBSException("E_AUTH_EXPIRED", "Credentials expired");
            }

            // Create save directory
            File saveFile = new File(params.savePath);
            File parentDir = saveFile.getParentFile();
            if (parentDir != null) {
                parentDir.mkdirs();
            }

            if (saveFile.exists()) {
                saveFile.delete();
            }

            // Send downloadStart event
            taskState.status = "DOWNLOADING";
            Map<String, Object> startEvent = new HashMap<>();
            startEvent.put("taskId", params.taskId);
            startEvent.put("objectKey", params.objectKey);
            eventEmitter.emit("downloadStart", startEvent);

            // Create download request
            GetObjectRequest request = new GetObjectRequest(params.bucket, params.objectKey);
            if (params.range != null) {
                try {
                    // Parse range header format: "bytes=START-END"
                    String rangeValue = params.range.replaceFirst("^bytes=", "").trim();
                    String[] parts = rangeValue.split("-", 2);
                    if (parts.length >= 1 && !parts[0].isEmpty()) {
                        request.setRangeStart(Long.parseLong(parts[0].trim()));
                    }
                    if (parts.length >= 2 && !parts[1].isEmpty()) {
                        request.setRangeEnd(Long.parseLong(parts[1].trim()));
                    }
                } catch (NumberFormatException e) {
                    Log.w(TAG, "[Download] invalid range format: " + params.range + ", ignoring");
                }
            }
            if (params.versionId != null) {
                request.setVersionId(params.versionId);
            }

            // Execute download
            Log.d(TAG, "[Download] sending GetObject request: bucket=" + params.bucket + "  key=" + params.objectKey);
            ObsObject obsObject = obsClientHolder.getClient().getObject(request);
            taskState.totalBytes = obsObject.getMetadata().getContentLength();
            taskState.etag = obsObject.getMetadata().getEtag() != null ?
                    obsObject.getMetadata().getEtag() : "";
            Log.i(TAG, "[Download] object size=" + taskState.totalBytes + " bytes  etag=" + taskState.etag);

            // Stream to file
            InputStream inputStream = obsObject.getObjectContent();
            FileOutputStream outputStream = new FileOutputStream(saveFile);

            try {
                byte[] buffer = new byte[BUFFER_SIZE];
                long lastProgressTime = System.currentTimeMillis();
                int bytesRead;

                while ((bytesRead = inputStream.read(buffer)) != -1) {
                    outputStream.write(buffer, 0, bytesRead);
                    long downloaded = taskState.downloadedBytes.addAndGet(bytesRead);

                    // Send progress event (with throttling)
                    long now = System.currentTimeMillis();
                    if (now - lastProgressTime >= PROGRESS_UPDATE_INTERVAL) {
                        emitDownloadProgress(taskState, downloaded, false);
                        lastProgressTime = now;
                    }
                }

                outputStream.flush();

                // Send final 100% progress
                emitDownloadProgress(taskState, taskState.downloadedBytes.get(), true);

            } finally {
                try {
                    inputStream.close();
                } catch (Exception e) {
                    // Ignore close errors
                }
                try {
                    outputStream.close();
                } catch (Exception e) {
                    // Ignore close errors
                }
            }

            // Download complete
            long duration = System.currentTimeMillis() - params.startTimeMs;
            double avgSpeed = duration > 0 ? Math.round(taskState.totalBytes * 1000.0 / duration * 100.0) / 100.0 : 0.0;

            DownloadResult result = new DownloadResult(
                    params.taskId,
                    params.objectKey,
                    params.savePath,
                    taskState.totalBytes,
                    taskState.etag,
                    duration,
                    avgSpeed
            );

            taskState.status = "COMPLETED";

            // Send success event
            Map<String, Object> successEvent = new HashMap<>();
            successEvent.put("taskId", params.taskId);
            successEvent.put("objectKey", result.objectKey);
            successEvent.put("savePath", result.savePath);
            successEvent.put("size", result.size);
            successEvent.put("etag", result.etag);
            successEvent.put("duration", result.duration);
            successEvent.put("avgSpeed", result.avgSpeed);
            eventEmitter.emit("downloadSuccess", successEvent);

            // Clear throttle record
            eventEmitter.clearThrottle(params.taskId);

            Log.i(TAG, "[Download] completed: objectKey=" + params.objectKey
                    + "  size=" + taskState.totalBytes + " bytes"
                    + "  duration=" + duration + "ms"
                    + "  speed=" + String.format("%.1f", avgSpeed / 1024) + " KB/s");

        } catch (Exception e) {
            throw ErrorMapper.mapException(e);
        }
    }

    /**
     * Emit download progress
     */
    private void emitDownloadProgress(DownloadTaskState taskState, long downloaded, boolean force) {
        int percentage = taskState.totalBytes > 0 ?
                (int)(downloaded * 100 / taskState.totalBytes) : 0;

        Map<String, Object> progressEvent = new HashMap<>();
        progressEvent.put("taskId", taskState.taskId);
        progressEvent.put("downloadedBytes", downloaded);
        progressEvent.put("totalBytes", taskState.totalBytes);
        progressEvent.put("percentage", percentage);

        eventEmitter.emitProgress(taskState.taskId, "downloadProgress", progressEvent, force);
    }

    /**
     * Handle download error
     */
    private void handleDownloadError(DownloadTaskState taskState, Throwable error) {
        OBSException obsError = ErrorMapper.mapException(error);
        taskState.error = obsError;
        taskState.status = "FAILED";
        Log.e(TAG, "[Download] failed: taskId=" + taskState.taskId + "  error=" + obsError.getMessage());

        // Send error event
        Map<String, Object> errorEvent = new HashMap<>();
        errorEvent.put("taskId", taskState.taskId);
        errorEvent.put("code", obsError.getCode());
        errorEvent.put("message", obsError.getMessage());
        errorEvent.put("isRetryable", obsError.isRetryable());
        eventEmitter.emit("downloadError", errorEvent);

        // Clear throttle record
        eventEmitter.clearThrottle(taskState.taskId);

        Log.e(TAG, "Download failed: " + taskState.taskId + ", error: " + obsError.getMessage());
    }

    /**
     * Cancel download
     */
    public void cancelDownload(String taskId) throws Exception {
        DownloadTaskState taskState = tasks.get(taskId);
        if (taskState == null) {
            Log.w(TAG, "cancelDownload: task not found (may have already completed): " + taskId);
            return;
        }

        // Cancel future
        if (taskState.future != null) {
            taskState.future.cancel(true);
        }

        taskState.status = "CANCELED";
        tasks.remove(taskId);

        // Send cancel event
        Map<String, Object> cancelEvent = new HashMap<>();
        cancelEvent.put("taskId", taskId);
        eventEmitter.emit("downloadCancel", cancelEvent);
    }

    /**
     * Cancel all downloads
     */
    public void cancelAll() {
        for (String taskId : tasks.keySet()) {
            try {
                cancelDownload(taskId);
            } catch (Exception e) {
                Log.w(TAG, "Failed to cancel task " + taskId + ": " + e.getMessage());
            }
        }
    }

    /**
     * Get task status
     */
    public Map<String, Object> getTaskStatus(String taskId) {
        DownloadTaskState taskState = tasks.get(taskId);
        if (taskState == null) {
            return null;
        }

        Map<String, Object> progress = new HashMap<>();
        progress.put("taskId", taskState.taskId);
        progress.put("downloadedBytes", taskState.downloadedBytes.get());
        progress.put("totalBytes", taskState.totalBytes);
        progress.put("percentage", taskState.totalBytes > 0 ?
                (int)(taskState.downloadedBytes.get() * 100 / taskState.totalBytes) : 0);

        Map<String, Object> status = new HashMap<>();
        status.put("taskId", taskState.taskId);
        status.put("type", "download");
        status.put("objectKey", taskState.params.objectKey);
        status.put("status", taskState.status);
        status.put("progress", progress);

        return status;
    }

    /**
     * Get all tasks
     */
    public List<Map<String, Object>> getAllTasks() {
        List<Map<String, Object>> result = new ArrayList<>();
        for (DownloadTaskState taskState : tasks.values()) {
            Map<String, Object> task = new HashMap<>();
            task.put("taskId", taskState.taskId);
            task.put("type", "download");
            task.put("objectKey", taskState.params.objectKey);
            task.put("status", taskState.status);
            result.add(task);
        }
        return result;
    }

    /**
     * Clear completed, failed, and canceled tasks
     */
    public void clearCompletedTasks() {
        tasks.entrySet().removeIf(entry -> {
            String status = entry.getValue().status;
            return "COMPLETED".equals(status) || "FAILED".equals(status) || "CANCELED".equals(status);
        });
    }

    /**
     * Cleanup resources
     */
    public void destroy() {
        executorService.shutdownNow();
        tasks.clear();
    }
}
